neqo-common/src/qlog.rs

```diff
@@ -117,15 +117,27 @@ impl Qlog {
 
     /// If logging enabled, closure is given the Qlog stream to write events and
     /// frames to.
-    pub fn add_event_with_stream<F>(&self, f: F)
+    pub fn add_event_with_stream<F>(&mut self, f: F)
     where
         F: FnOnce(&mut QlogStreamer) -> Result<(), qlog::Error>,
     {
-        if let Some(inner) = self.inner.borrow_mut().as_mut() {
-            if let Err(e) = f(&mut inner.streamer) {
-                log::error!("Qlog event generation failed with error {e}; closing qlog.");
-                *self.inner.borrow_mut() = None;
-            }
+        let Some(inner) = self.inner.as_mut() else {
+            return;
+        };
+
+        let mut borrow = inner.borrow_mut();
+
+        let Some(inner_inner) = borrow.as_mut() else {
+            return;
```

Comment:
You *could* do this here, which would mean that subsequent calls would hit the first check:

```suggestion
            self.inner = None;
            return;
```

... At least as far as the clone that this is operating on.  Other clones will do the same thing when they first touch a disabled log.

After all, the point of using `RefCell` is to avoid needing a mutable reference.  But since you made the reference mutable, we have the luxury of updating this.

The drawback is that this is a ratchet.  If the inner is ever set to `None`, it can never be set back to `Some`, because any clone that calls this function will disconnect from the `Rc`.  That means that you need to ensure that you initialize logging before using the connection.  If you don't, you can't enable.